ydb_dbapi/connections.py

```diff
@@ -0,0 +1,351 @@
+
+import posixpath
+from typing import NamedTuple
+from typing import TypedDict
+
+import ydb
+from typing_extensions import NotRequired
+from typing_extensions import Unpack
+from ydb.retries import retry_operation_async
+from ydb.retries import retry_operation_sync
+
+from .cursors import AsyncCursor
+from .cursors import Cursor
+from .errors import InterfaceError
+from .errors import InternalError
+from .errors import NotSupportedError
+from .utils import handle_ydb_errors
+
+
+class IsolationLevel:
+    SERIALIZABLE = "SERIALIZABLE"
+    ONLINE_READONLY = "ONLINE READONLY"
+    ONLINE_READONLY_INCONSISTENT = "ONLINE READONLY INCONSISTENT"
+    STALE_READONLY = "STALE READONLY"
+    SNAPSHOT_READONLY = "SNAPSHOT READONLY"
+    AUTOCOMMIT = "AUTOCOMMIT"
+
+
+class ConnectionKwargs(TypedDict):
+    credentials: NotRequired[ydb.AbstractCredentials]
+    ydb_table_path_prefix: NotRequired[str]
+    ydb_session_pool: NotRequired[
+        ydb.QuerySessionPool | ydb.aio.QuerySessionPool
+    ]
+
+
+class BaseConnection:
+    _tx_mode: ydb.BaseQueryTxMode = ydb.QuerySerializableReadWrite()
+    _tx_context: ydb.QueryTxContext | ydb.aio.QueryTxContext | None = None
+    interactive_transaction: bool = False
+    _shared_session_pool: bool = False
+
+    _driver_cls = ydb.Driver
+    _pool_cls = ydb.QuerySessionPool
+    _cursor_cls: type[Cursor | AsyncCursor] = Cursor
+
+    _driver: ydb.Driver | ydb.aio.Driver
+    _pool: ydb.QuerySessionPool | ydb.aio.QuerySessionPool
+
+    _current_cursor: AsyncCursor | Cursor | None = None
+
+    def __init__(
+        self,
+        host: str = "",
+        port: str = "",
+        database: str = "",
+        **conn_kwargs: Unpack[ConnectionKwargs],
+    ) -> None:
+        self.endpoint = f"grpc://{host}:{port}"
+        self.database = database
+        self.conn_kwargs = conn_kwargs
+        self.credentials = self.conn_kwargs.pop("credentials", None)
+        self.table_path_prefix = self.conn_kwargs.pop(
+            "ydb_table_path_prefix", ""
+        )
+
+        if (
+            "ydb_session_pool" in self.conn_kwargs
+        ):  # Use session pool managed manually
+            self._shared_session_pool = True
+            self._session_pool = self.conn_kwargs.pop("ydb_session_pool")
+            self._driver = self._session_pool._driver
+        else:
+            driver_config = ydb.DriverConfig(
+                endpoint=self.endpoint,
+                database=self.database,
+                credentials=self.credentials,
+            )
+            self._driver = self._driver_cls(driver_config)
+            self._session_pool = self._pool_cls(self._driver, size=5)
+
+        self._session: ydb.QuerySession | ydb.aio.QuerySession | None = None
+
+    def set_isolation_level(self, isolation_level: str) -> None:
+        class IsolationSettings(NamedTuple):
+            ydb_mode: ydb.BaseQueryTxMode
+            interactive: bool
+
+        ydb_isolation_settings_map = {
```

Comment:
**suggestion**: Let's move this to the global variable to avoid recreation of dict on each call